### PR TITLE
Add support for etcd over TLS

### PR DESCRIPTION
**Pull request by gdb**
Are you interested in this functionality? I've checked that Vault did not work with my TLS-secured etcd setup without this patch, and that it works with it.

I'm happy to add docs and/or tests as appropriate.

### CODE REVIEW

File: physical/etcd.go [hunk 1]
```diff
@@ -79,10 +79,23 @@ func newEtcdBackend(conf map[string]string) (Backend, error) {
...
 
-	// Create a new client from the supplied addres and attempt to sync with the
+	// Create a new client from the supplied address and attempt to sync with the
 	// cluster.
-	client := etcd.NewClient(strings.Split(machines, EtcdMachineDelimiter))
+	var client *etcd.Client
+	cert, has_cert := conf["tls_cert_file"]
+	key, has_key := conf["tls_key_file"]
+	ca, has_ca := conf["tls_ca_file"]
+	if has_cert || has_key || has_ca {
```
Comment (jefferai) [inline physical/etcd.go]: This probably needs to check for `(has_cert && has_key) || has_ca`; if you have a client cert but no key, or vice versa, you can't actually auth.

**Comment by jefferai** [comment]
I have an ask for you, if you're up for it: since you're actively using etcd and can properly test it, any chance you can look at porting from the deprecated `go-etcd` to the new official client (https://godoc.org/github.com/coreos/etcd/client)? It probably wouldn't be too difficult, but I do know that there are some API changes, including in the code you put here.

**Comment by gdb** [comment]
Yep, no problem. (May be a bit before I get around to it though.)

**Comment by jefferai** [comment]
We have a 0.4 release coming soon (a week or two). If you think you can get to it before then that would be great. Otherwise we can merge this in, and maybe you'll be motivated to do that at some future point :-)

**Comment by gdb** [comment]
Yep, that should be doable.

**Comment by jefferai** [comment]
Great. I'll hold off on this for now (but see my line comment above -- regardless of which library is being used, probably want to ensure that a client cert/key are always provided together).

**Comment by jefferai** [comment]
@gdb Any idea on when you might be able to work on this? We're hoping for an RC early next week.

Alternately, if it's easy, I could work on fixing this in code, and then you could do the testing/QA. That's the hard part for me. If you think you could easily do that, I'll make the code changes.

**Comment by vicki-c** [comment]
@jefferai I can probably work on it this weekend.

**Comment by jefferai** [comment]
@vicki-c @gdb Great!

I also wanted to make you aware of something else -- there is now a basic connection token pool in the `physical` package. You can see how it's used in the `consul` and `inmem` backends -- it's quite simple, but it would be a good idea to add that if possible. We've seen a couple cases with huge numbers of revocations (such as ones that are past-due when someone had Vault off for a while and then opened it back up) causing huge numbers of connections to the storage backend, so using the tokens to control access keeps that from hammering the storage backend.

**Comment by gdb** [comment]
(Closing in favor of @vicki-c's new PR #780.)

**Pull request closed without merging**